host/src/proof.rs

```diff
@@ -384,6 +388,90 @@ impl ProofActor {
 
         Ok(())
     }
+
+    async fn cancel_all_running_tasks(&mut self) -> HostResult<()> {
+        info!("Cancelling all running tasks");
+
+        // Clone all tasks to avoid holding locks to avoid deadlock, they will be locked by other
+        // internal functions.
+        let running_tasks = {
+            let running_tasks = self.running_tasks.lock().await;
+            (*running_tasks).clone()
+        };
+
+        // Cancel all running tasks, don't stop even if any task fails.
+        let mut final_result = Ok(());
+        for proof_task_descriptor in running_tasks.keys() {
+            match self.cancel_task(proof_task_descriptor.clone()).await {
+                Ok(()) => {
+                    info!(
+                        "Cancel task during system pause, task: {:?}",
+                        proof_task_descriptor
+                    );
+                }
+                Err(e) => {
+                    error!(
+                        "Failed to cancel task during system pause: {}, task: {:?}",
+                        e, proof_task_descriptor
+                    );
+                    final_result = final_result.and(Err(e));
+                }
+            }
+        }
+        final_result
+    }
+
+    async fn cancel_all_aggregation_tasks(&mut self) -> HostResult<()> {
+        info!("Cancelling all aggregation tasks");
+
+        // Clone all tasks to avoid holding locks to avoid deadlock, they will be locked by other
+        // internal functions.
+        let aggregate_tasks = {
+            let aggregate_tasks = self.aggregate_tasks.lock().await;
+            (*aggregate_tasks).clone()
+        };
+
+        // Cancel all aggregation tasks, don't stop even if any task fails.
+        let mut final_result = Ok(());
+        for request in aggregate_tasks.keys() {
+            match self.cancel_aggregation_task(request.clone()).await {
+                Ok(()) => {
+                    info!(
+                        "Cancel aggregation task during system pause, task: {}",
+                        request
+                    );
+                }
+                Err(e) => {
+                    error!(
+                        "Failed to cancel aggregation task during system pause: {}, task: {}",
+                        e, request
+                    );
+                    final_result = final_result.and(Err(e));
+                }
+            }
+        }
+        final_result
+    }
+
+    async fn handle_system_pause(&mut self) -> HostResult<()> {
+        info!("System pausing");
+
+        let mut final_result = Ok(());
+
+        self.pending_tasks.lock().await.clear();
+
+        if let Err(e) = self.cancel_all_running_tasks().await {
```

Comment:
Do we need a flag to indicate to the system how to handle running tasks? possible choices should include cancel, wait, and pause(will be pickup by others or a restarted self, the end-game design).

Reply:
Sounds good. Would you like to provide some more design detail? I don't have a detailed design yet, thank you.

Reply:
no valid one so far.
for now, sp1 does not support cancel but r0 does, I just guess maybe we need to give each prover different way(on their own) to handle this pause, also need consider the future pick-up (resume) approach. can discuss it later.